### DISCUSSION TITLE
New release 0.13

### DISCUSSION THREAD

**Discussion by jghauser** [papis/papis#549] [General]
What do you think about releasing a new version? I've been more active on my neovim plugin recently and was thinking it would be nice if it could be used with a released version of papis. Quite a lot has happened since the last release, and it's been (almost) a year.

**Comment by alexfikl** [comment]
Totally, I'm all for it! 

I don't think there's anything holding it back (?), but we do need @alejandrogallo to make some tags and push some things to PyPI (not quite sure what's involved in the release :))

**Reply by jghauser** [reply]
ok great, let's see what @alejandrogallo thinks :)

**Comment by alejandrogallo** [comment]
I'm sorry I was absent lately!
Yes, a new release is totally needed, I already fixed some things in the web app for the last time we wanted to do a release, so I think this is it. I think the changelog is also quite complete, as I think @alexfikl gave it some love?

**Reply by alexfikl** [reply]
No worries at all! Hope you're doing well!

Yeah, the changelog should be in a reasonable shape, so feel free to merge #506 and make a release whenever :rocket:

**Reply by jghauser** [reply]
Excellent, looking forward to this! :)

**Comment by alejandrogallo** [comment]
I just updated it, thank you so much you two, papis would not be possible without your amazing engagement!

**Reply by alexfikl** [reply]
Woo! It's been a lot of fun! :tada:

**Comment by pmiam** [comment]
Hello! I would like to ask for your opinions on a PR I am considering:

I just started running the 0.13 release. The new "papis_id" index fulfills my basic requirements, but I noticed some rough spots I think I could smooth over.
1. document hashing is done in multiple places and should be consolidated
   - `compute_an_id` function in id.py produces the keys for info files 
   - `get_document_hash_folder` function in paths.py produces keys for the default add-folder-name

`compute_an_id` is simpler and also better integrated into the info.yaml maintenance/doctor functions, so it should stay. If it can be run before format-time -- i.e. right before the ref construction as of #484 -- then the resulting value would be available in templates.

2. early papis_id availability would thus simplify the default add-folder-name. Users concerned about coincident hashes could simply set it to {doc[time-added]}{doc[papis_id]} as well. Either would be prettier and more consistent than the current hash-author default.

I prefer to interact with most of my libraries through the UI. So, flat machine-managed directory trees are ideal for me. At least one other user #364 wants doc[papis_id] available at format time. 

The aesthetic improvement alone motivates me to contribute, but I'd like to clear up:
1. Is id generation defined in multiple places intentionally?
2. would this be a breaking change for anybody? ( I think not? )
3. anything I missed?

If not, I'll get on this immediately. Thank you all for this wonderful tool.

**Reply by alexfikl** [reply]
This sounds like a good idea to me! There are quite a few changes since 0.13 (we should be releasing 0.14 soon-ish), but I think all of your points stand.

> Is id generation defined in multiple places intentionally?

No, at least as far as I know. It's purely historical: the folder hashing was introduced way before the `papis_id` and was never reconsidered.

> would this be a breaking change for anybody? ( I think not? )

I don't think so either. Both `compute_an_id` and `get_document_hash_folder` use MD5 hashes, so as far as any user is concerned, they should be equivalent.

> anything I missed?

The document `papis_id` is generally computed in `papis.commands.add` at the end of the whole process. This means that it is computed using the the document metadata as it will be added to the library (i.e. saved to its `info.yaml` file). By computing the `papis_id` before all that, it's no longer based on the actual data being saved..

Not sure that's any sort of an issue, since at the end of the day we just use the ID as a unique identifier, not to compare hashes between documents (e.g. for integrity checking), so it should be just fine.

Also, feel free to open a new issue or a separate discussion if you want to talk about this in more details!

**Reply by choreutes** [reply]
For a second I was thinking that maybe one might want to use the IDs for integrity or similar later on and that we should think about this before changing order here. But then I realized that users may change information about documents after they have already been added to the library and the ID is not recalculated in this case. So IDs are unsuited for integrity checking anyway. So I guess this is probably fine.

Also the crypto person in me is kind of screaming that one shouldn't use MD5 because it is not secure. In our use-case document IDs are probably not security relevant (although colliding document IDs may cause problems?) so I guess this is also fine. I have learned though that some exotic python distributions disable md5 in hashlib because of the security issues and Papis would break in such environments.

**Reply by alexfikl** [reply]
> Also the crypto person in me is kind of screaming that one shouldn't use MD5 because it is not secure. In our use-case document IDs are probably not security relevant (although colliding document IDs may cause problems?) so I guess this is also fine.

Yeah, we're just using it as a "give me a length 32 random looking string" function, so I don't think it should have any security implications. But MD5 is prone to collisions (at least comparatively), so we are doing the naive thing and keep generating them until it doesn't collide!
https://github.com/papis/papis/blob/8c0196981242e4b7342ee87fc9088c13df8f80cd/papis/database/base.py#L108-L125

> I have learned though that some exotic python distributions disable md5 in hashlib because of the security issues and Papis would break in such environments.

Hm.. that's good to know! Thank you!

**Comment by ExpandingMan** [comment]
Any plans for an 0.14 release?

**Reply by jghauser** [reply]
Yes, see #799. :)
